src/sqlfluff/dialects/dialect_snowflake.py

```diff
@@ -682,6 +684,171 @@
+    """
+
+    type = "object_reference"
+    match_grammar: Matchable = OneOf(
+        # Snowflake's IDENTIFIER pseudo-function
+        # https://docs.snowflake.com/en/sql-reference/identifier-literal.html
+        Delimited(
+            Sequence(
+                "IDENTIFIER",
+                Bracketed(
+                    OneOf(
+                        Ref("SingleQuotedIdentifierSegment"),
+                        Ref("ReferencedVariableNameSegment"),
+                    ),
+                ),
+            ),
+        ),
+        # Copied from ansi.ObjectReferenceSegment
+        # match grammar (don't allow whitespace)
+        Delimited(
+            Ref("SingleIdentifierGrammar"),
+            delimiter=OneOf(
+                Ref("DotSegment"), Sequence(Ref("DotSegment"), Ref("DotSegment"))
+            ),
+            terminator=OneOf(
+                "ON",
+                "AS",
+                "USING",
+                Ref("CommaSegment"),
+                Ref("CastOperatorSegment"),
+                Ref("StartSquareBracketSegment"),
+                Ref("StartBracketSegment"),
+                Ref("BinaryOperatorGrammar"),
+                Ref("ColonSegment"),
+                Ref("DelimiterGrammar"),
+                Ref("JoinLikeClauseGrammar"),
+                BracketedSegment,
+            ),
+            allow_gaps=False,
+        ),
+    )
+
+
+class TableReferenceSegment(ObjectReferenceSegment):
+    """A reference to an table, CTE, subquery or alias.
+
+    Overriding to capture Snowflakes's override of ObjectReferenceSegment
+    """
+
+    type = "table_reference"
+
+
+class SchemaReferenceSegment(ObjectReferenceSegment):
+    """A reference to a schema.
+
+    Overriding to capture Snowflakes's override of ObjectReferenceSegment
+    """
+
+    type = "schema_reference"
+
+
+class DatabaseReferenceSegment(ObjectReferenceSegment):
+    """A reference to a database.
+
+    Overriding to capture Snowflakes's override of ObjectReferenceSegment
+    """
+
+    type = "database_reference"
+
+
+class IndexReferenceSegment(ObjectReferenceSegment):
+    """A reference to an index.
+
+    Overriding to capture Snowflakes's override of ObjectReferenceSegment
+    """
+
+    type = "index_reference"
+
+
+class ExtensionReferenceSegment(ObjectReferenceSegment):
+    """A reference to an extension.
+
+    Overriding to capture Snowflakes's override of ObjectReferenceSegment
+    """
+
+    type = "extension_reference"
+
+
+class ColumnReferenceSegment(ObjectReferenceSegment):
+    """A reference to column, field or alias.
+
+    Overriding to capture Snowflakes's override of ObjectReferenceSegment
+    """
+
+    type = "column_reference"
+
+
+class SequenceReferenceSegment(ObjectReferenceSegment):
+    """A reference to a sequence.
+
+    Overriding to capture Snowflakes's override of ObjectReferenceSegment
+    """
+
+    type = "sequence_reference"
+
+
```

Comment:
Do you really need these? They inherit from `ObjectReferenceSegment` so would have though they would automatically have picked up the changes to that above?

```suggestion
```

Or did you try that, and it didn't work?

Reply:
Holy cow. `SingleIdentifierGrammar` seems much cleaner. Good idea.

Reply:
As you'll see in the updated version, I removed nearly all of those extra classes once the `IDENTIFIER` syntax moved to `SingleIdentifierGrammar`. But `FunctionNameSegment` seems to be distinct. I couldn't find a way around keeping that one.